### ISSUE TITLE
dpd_ssl_server signature does not trigger

### ISSUE THREAD

**Issue by fox-ds** [zeek/zeek#1952]
In the following pcap (part 2) there is TLS1.2 traffic on port 787.
  
https://www.malware-traffic-analysis.net/2022/01/27/index.html

Using Zeek 4.2.0 only the first two connections are identified as SSL

The dpd.sig contains a client and server signature. I could validate that the client signature triggers, however I cant get the server signature to work. Changing it to match on a wildcard doen't even trigger:
```
signature dpd_ssl_server {
  ip-proto == tcp
  # Server hello.
  payload /.*/
  requires-reverse-signature dpd_ssl_client
  enable "ssl"
  tcp-state responder
}
```
The issue: why doesn't the server signature trigger and how big is the performance impact if the SSL analyzer is enabled on the client-hello?

Running 1 connection with debug flags shows that the server hello is sent to the dpd analyzer.

```
1643372001.773339/1643722607.313178 [dpd] TCP added packet child CONNSIZE
1643372001.773339/1643722607.313187 [dpd] TCP[1] added child PIA_TCP[2]
1643372001.773339/1643722607.313271 [dpd] PIA_TCP[2] FirstPacket(T)
1643372001.773339/1643722607.313358 [dpd] CONNSIZE[3] DeliverPacket(0, T, 1, 0x6d485e0, 0) []
1643372001.854203/1643722607.313417 [dpd] PIA_TCP[2] FirstPacket(F)
1643372001.854203/1643722607.313425 [dpd] CONNSIZE[3] DeliverPacket(0, F, 1, 0x6d485e0, 0) []
1643372001.854574/1643722607.313452 [dpd] CONNSIZE[3] DeliverPacket(0, T, 1, 0x6d485e0, 0) []

CLIENT HELLO
1643372001.854988/1643722607.313470 [dpd] PIA_TCP[2] DeliverStream(1460, T) [\x16\x03\x03\x06\xf5\x01\x00\x06\xf1\x03\x03a\xf3\xdd\xe1\xf1\x05\xa5\x14b\xdc\xderZ\xe25\xf9\xc8%,*\xe8\xa7Dw8\x93\xa9\xfcL...]
1643372001.854988/1643722607.313919 [dpd] CONNSIZE[3] DeliverPacket(1460, T, 1, 0x6d485e0, 1460) [\x16\x03\x03\x06\xf5\x01\x00\x06\xf1\x03\x03a\xf3\xdd\xe1\xf1\x05\xa5\x14b\xdc\xderZ\xe25\xf9\xc8%,*\xe8\xa7Dw8\x93\xa9\xfcL...]
1643372001.855002/1643722607.313935 [dpd] PIA_TCP[2] DeliverStream(326, T) [\x8a\xa8\xa4\xe2\x0d(\xb7\xd8\xa5\x1a02\xa4\T\xa7@J\x12X'{.\xf8X\x01\x82\x15\x93\xa3\xf8\x0aW4\xd49\xb4\xee\xce\x94...]
1643372001.855002/1643722607.313941 [dpd] CONNSIZE[3] DeliverPacket(326, T, 1461, 0x6d485e0, 326) [\x8a\xa8\xa4\xe2\x0d(\xb7\xd8\xa5\x1a02\xa4\T\xa7@J\x12X'{.\xf8X\x01\x82\x15\x93\xa3\xf8\x0aW4\xd49\xb4\xee\xce\x94...]
1643372001.855074/1643722607.313950 [dpd] CONNSIZE[3] DeliverPacket(0, F, 1, 0x6d485e0, 0) []
1643372001.855100/1643722607.313958 [dpd] CONNSIZE[3] DeliverPacket(0, F, 1, 0x6d485e0, 0) []

SERVER HELLO
1643372001.984122/1643722607.313968 [dpd] PIA_TCP[2] DeliverStream(94, F) **[\x16\x03\x03\x00Y\x02\x00\x00U\x03\x03!k\xa9\x1d\xe2NZ\x14\xb9K@\x81\xed\xaeP\xdfB\xa0\xec\xd7i\x81H\xafDOWNG.**..]
1643372001.984122/1643722607.313973 [dpd] CONNSIZE[3] DeliverPacket(94, F, 1, 0x6d485e0, 94) [\x16\x03\x03\x00Y\x02\x00\x00U\x03\x03!k\xa9\x1d\xe2NZ\x14\xb9K@\x81\xed\xaeP\xdfB\xa0\xec\xd7i\x81H\xafDOWNG...]
1643372001.984408/1643722607.313981 [dpd] CONNSIZE[3] DeliverPacket(0, T, 1787, 0x6d485e0, 0) []

APPLICATION DATA
1643372001.985646/1643722607.313992 [dpd] PIA_TCP[2] DeliverStream(1346, F) [\x16\x03\x03\x065\x04\x00\x061\x00\x01Q\x80\x06+|\xeb\xd6\xc1\x03\xc4\xfa\xae\xeb\xe7iWAIg1t\x81\xa2\xf2\xd5\x99+\xc2\x08...]
1643372001.985646/1643722607.313998 [dpd] CONNSIZE[3] DeliverPacket(1346, F, 95, 0x6d485e0, 1346) [\x16\x03\x03\x065\x04\x00\x061\x00\x01Q\x80\x06+|\xeb\xd6\xc1\x03\xc4\xfa\xae\xeb\xe7iWAIg1t\x81\xa2\xf2\xd5\x99+\xc2\x08...]
1643372001.985879/1643722607.314006 [dpd] CONNSIZE[3] DeliverPacket(0, T, 1787, 0x6d485e0, 0) []
1643372002.054773/1643722607.314016 [dpd] PIA_TCP[2] DeliverStream(299, F) [~\xf7E*\xdb\x86\x19x\xb2\xc2\xce\x94\x04"\xef^sr4\xc8\xea\xb1\xf5\xa5\xed[L\x1e\xcb<\x9d\xcb\xa7n\xeah\x1f\xc8\x10\x85...]
1643372002.054773/1643722607.314021 [dpd] CONNSIZE[3] DeliverPacket(299, F, 1441, 0x6d485e0, 299) [~\xf7E*\xdb\x86\x19x\xb2\xc2\xce\x94\x04"\xef^sr4\xc8\xea\xb1\xf5\xa5\xed[L\x1e\xcb<\x9d\xcb\xa7n\xeah\x1f\xc8\x10\x85...]
1643372002.055192/1643722607.314029 [dpd] CONNSIZE[3] DeliverPacket(0, T, 1787, 0x6d485e0, 0) []
1643372002.055783/1643722607.314039 [dpd] PIA_TCP[2] DeliverStream(51, T) [\x14\x03\x03\x00\x01\x01\x16\x03\x03\x00(\x00\x00\x00\x00\x00\x00\x00\x00m\xe8E\xc2X\xc4\x94&e\x9d8|U>\xaf\x17\xb4=\x1b\xdd\xc4...]
1643372002.055783/1643722607.314044 [dpd] CONNSIZE[3] DeliverPacket(51, T, 1787, 0x6d485e0, 51) [\x14\x03\x03\x00\x01\x01\x16\x03\x03\x00(\x00\x00\x00\x00\x00\x00\x00\x00m\xe8E\xc2X\xc4\x94&e\x9d8|U>\xaf\x17\xb4=\x1b\xdd\xc4...]
1643372002.055994/1643722607.314055 [dpd] CONNSIZE[3] DeliverPacket(0, F, 1740, 0x6d485e0, 0) []
1643372002.057618/1643722607.314067 [dpd] PIA_TCP[2] DeliverStream(658, T) [\x17\x03\x03\x02\x8d\x00\x00\x00\x00\x00\x00\x00\x01\x1f\xfc\x94\xc2\x08@Y\x93p\xb6\x04Y\x87\xdf\x8d\xeb\xfdzK\xb9\xcc\x9e\xf9\xd2'\xf8'...]
1643372002.057618/1643722607.314073 [dpd] CONNSIZE[3] DeliverPacket(658, T, 1838, 0x6d485e0, 658) [\x17\x03\x03\x02\x8d\x00\x00\x00\x00\x00\x00\x00\x01\x1f\xfc\x94\xc2\x08@Y\x93p\xb6\x04Y\x87\xdf\x8d\xeb\xfdzK\xb9\xcc\x9e\xf9\xd2'\xf8'...]
1643372002.057714/1643722607.314081 [dpd] CONNSIZE[3] DeliverPacket(0, F, 1740, 0x6d485e0, 0) []
1643372002.252251/1643722607.314091 [dpd] PIA_TCP[2] DeliverStream(197, F) [\x17\x03\x03\x00\xa1\x00\x00\x00\x00\x00\x00\x00\x01\xf6\xcaI\x0e\x12\xf6\xbf0\xa3\xd6\\xe9\xbe\x91\x7f\x06%\x1b\xc7\x8f\xbbn\xfc\x0b:\x00i...]
1643372002.252251/1643722607.314098 [dpd] CONNSIZE[3] DeliverPacket(197, F, 1740, 0x6d485e0, 197) [\x17\x03\x03\x00\xa1\x00\x00\x00\x00\x00\x00\x00\x01\xf6\xcaI\x0e\x12\xf6\xbf0\xa3\xd6\\xe9\xbe\x91\x7f\x06%\x1b\xc7\x8f\xbbn\xfc\x0b:\x00i...]
1643372002.252640/1643722607.314106 [dpd] CONNSIZE[3] DeliverPacket(0, T, 2496, 0x6d485e0, 0) []
1643372002.252813/1643722607.314116 [dpd] CONNSIZE[3] DeliverPacket(0, T, 2496, 0x6d485e0, 0) []
1643372002.252859/1643722607.314125 [dpd] CONNSIZE[3] DeliverPacket(0, F, 1938, 0x6d485e0, 0) []
1643372002.252859/1643722607.314208 [dpd] TCP[1] EndOfData(T)
1643372002.252859/1643722607.314213 [dpd] TCP[1] EndOfData(F)
1643372002.252859/1643722607.314216 [dpd] PIA_TCP[2] EndOfData(T)
1643372002.252859/1643722607.314219 [dpd] PIA_TCP[2] EndOfData(F)
1643372002.252859/1643722607.314230 [dpd] CONNSIZE[3] EndOfData(T)
1643372002.252859/1643722607.314233 [dpd] CONNSIZE[3] EndOfData(F)
```

**Comment by 0xxon** [comment]
Hi,

thank you for this report - that pcap is interesting. The problem is that the TLS handshake in these examples is quite large - the client hello is often larger than 1kb. This is larger than the default `dpd_buffer_size` of Zeek. In the default configuration, this means that the protocol analyzer will no longer be enabled, since the packet data cannot be replayed to it anymore (`dpd_match_only_beginning` is enabled by default).

This is obviously not optimal. As a fix, you can just redefine `dpd_buffer_size` to be larger - a script that contains just `redef dpd_buffer_size = 5120;` will make this work.

But - since we encountered this before (for HTTP), and since this means that SSL handshakes will now probably often be larger than the 1k limit we have so far - I think it might be best for us to change this setting to a larger value by default.

@rsmmr / @JustinAzoff might have some opinions on this too.

**Comment by JustinAzoff** [comment]
I had tried the 'just make dpd buffer size larger' approach for http, but that still failed in some cases.  There was always a reason for something to send an even larger request.  We finally just went with the change to make the dpd sig not require the reverse match.

That is probably the correct approach here too.. if the client does a tls handshake, we can attach the tls analyzer and see what happens. Worse case we should be generating the events for the client side of the handshake and then the server will reply with something non-tls and it will trigger a ProtocolViolation.
